### PR TITLE
[stable/prisma]: add Prisma chart

### PR DESCRIPTION
**Pull request by giacope**
**What this PR does / why we need it**:
This PR adds a chart for installing [Prisma](https://www.prisma.io/) on Kubernetes.
Prisma is a performant open-source GraphQL ORM-like layer doing the heavy lifting in your GraphQL server. It helps you easily build a GraphQL server that uses any database.

### CODE REVIEW

File: incubator/prisma/templates/configmap.yaml
Comment (jainishshah17) [inline incubator/prisma/templates/configmap.yaml]: @giacomoguiulfo I recommend converting this to secret.

File: incubator/prisma/values.yaml
Comment (jainishshah17) [inline incubator/prisma/values.yaml]: @giacomoguiulfo avoid hardcoding password.

**Comment by YaraMohammed** [comment]
I tried to install the chart and access prisma pod using port forward but I got this error not sure if I missed something? 
screen shot 2018-07-06 at 3 07 56 pm

**Comment by giacomoguiulfo** [comment]
@YaraMohammed What do the logs for Prisma say?

File: incubator/prisma/templates/configmap.yaml
Comment (giacope) [reply incubator/prisma/templates/configmap.yaml]: Thanks for the review!
I don't think the full config should be in a secret. But definitely the auth secret should be in a secret. I will work on this. I hope Prisma supports passing the auth secret in a different way.

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: Thanks! Ok, I will work on this.

**Comment by YaraMohammed** [comment]
@giacomoguiulfo  
```
Obtaining exclusive agent lock...
Obtaining exclusive agent lock... Successful.
Deployment worker initialization complete.
Initializing workers...
Successfully started 1 workers.
Server running on :4466
Version is up to date.
[Metrics] Warning: no metrics will be recorded.
```

**Comment by giacomoguiulfo** [comment]
@YaraMohammed can you reach the server through other means, like `curl` or `nc`?

File: incubator/prisma/templates/deployment.yaml
Comment (YaraMohammed) [inline incubator/prisma/templates/deployment.yaml]: port name is defined as `prisma` in the deployment but in the service `targetPort: http`

Comment (giacope) [reply incubator/prisma/templates/deployment.yaml]: Good catch 👍

**Comment by emmenko** [comment]
> I tried to install the chart and access prisma pod using port forward but I got this error Server cannot be reached not sure if I missed something?

@YaraMohammed The error most likely means that you need to deploy the SDL schema first (https://www.prisma.io/docs/1.12/tutorials/deploy-prisma-servers/kubernetes-aiqu8ahgha#configuration-of-the-prisma-cli).

File: incubator/prisma/templates/configmap.yaml
Comment (emmenko) [reply incubator/prisma/templates/configmap.yaml]: Yeah, this definitely needs to be defined as a `Secret`. I'm actually surprised that the documentation does not mention this (https://www.prisma.io/docs/1.12/tutorials/deploy-prisma-servers/kubernetes-aiqu8ahgha).

In my chart I put the entire `PRISMA_CONFIG` into a secret.

That's the only way that you can currently use this (https://github.com/prismagraphql/prisma/blob/665eb16138dfea9082746e21743121eccfc41419/server/libs/prisma-config/src/main/scala/com/prisma/config/ConfigLoader.scala#L45-L52). It would make sense to be able to pass multiple ENV variables.

/cc @akoenig @nikolasburk

File: incubator/prisma/values.yaml
Comment (emmenko) [inline incubator/prisma/values.yaml]: I think the default should be `false`. But good idea to install the chart automatically!

File: incubator/prisma/values.yaml
Comment (emmenko) [inline incubator/prisma/values.yaml]: We should also be able to specify the image tag

File: incubator/prisma/values.yaml
Comment (emmenko) [inline incubator/prisma/values.yaml]: Is the idea to also support other connectors like mysql?

**Review by emmenko — COMMENTED** [review]
Thanks for getting this started! 🙌

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: It should be possible to configure other connectors. In order to do so, disable `postgresql` and then set proper configs in the the `database:` section. However, we could also add a MySQL chart as another conditionally dependency, to simplify configurations. What do you think? @emmenko

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: All PostgreSQL chart values can be configured from this file, including the image tag; you just need to add the field. That said, I'm thinking maybe it should be visible from this particular `values.yaml` file due to its importance.

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: According to the technical requirements of the contribution guidelines, the pods should enter a running state with a `helm install` using the default values. If a database connector is not available, Prisma will be in a crash loop state. That's why we need a connector deployed by default.

File: incubator/prisma/templates/configmap.yaml
Comment (giacope) [reply incubator/prisma/templates/configmap.yaml]: Thanks for the review!
Yes, we could have the full config in a secret (for now). However, do you have any suggestion on how to approach this? The templating can get really uggly since it has to be base64 encoded and it is particularly long for a secret.

Comment (emmenko) [reply incubator/prisma/templates/configmap.yaml]: That's pretty simple to do. I usually have a `secret.yaml` file that describes the `Secret` resource, reads up the secrets from the "values" and encodes them into the `Secret`.

```yaml
apiVersion: v1
kind: Secret
metadata:
  # The identifier of this Secret, unique within a namespace.
  # https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
  name: {{ .Chart.Name }}-secret
  # Labels are custom fields useful for querying and filtering resources.
  # https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
  labels:
    app: {{ .Chart.Name }}
    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    component: {{ .Values.componentName }}
    heritage: {{ .Release.Service }}
    release: {{ .Release.Name }}
type: Opaque
data:
  # Inject (sensitive) environment variables from `Values` config.
  {{- range $key, $val := .Values.secrets }}
  {{ $key }}: {{ $val | b64enc }}
  {{- end}}
```

This assumes that you have a `secrets` object map in the `values.yaml`.

```yaml
# values/<env>/secrets.yaml
secrets:
  PRISMA_MANAGEMENT_API_SECRET: xxx
  PRISMA_CONFIG: |
    port: 4466
    managementApiSecret: xxx
    databases:
      default:
        connector: postgres
        host: localhost
        port: 5432
        user: postgres
        password: 123
        migrations: true
```

Because of the ability to "merge" values together, I usually have a `values.yaml` and a `secrets.yaml` which I pass to helm when installing

```bash
$ helm upgrade --install <release-name> \
  --namespace <namespace> \
  -f values/<env>/values.yaml \
  -f values/<env>/secrets.yaml \
  <chart-name>
```

Now, since the "secrets" should be encrypted, I use Google Cloud KMS to encrypt the `secrets.yaml` file into a `secrets.enc` and then decrypt the secrets on CI before upgrading the chart

```bash
$ gcloud kms decrypt \
  --project=<project-id> \
  --location=global \
  --keyring=<key-ring> \
  --key=<crypto-key> \
  --plaintext-file=values/<env>/secrets.yaml \
  --ciphertext-file=values/<enc>/secrets.enc
```

Finally, the secret values can be then referenced from the `Deployment`

```yaml
spec:
  template:
    spec:
      containers:
      - image: "prismagraphql/prisma:1.12"
        # Inject (non sensitive) environment variables from the `Values` config.
        env: {{- range $key, $val := .Values.config }}
        - name: {{ $key }}
          value: {{ $val | quote }}
          {{- end}}
        # Inject (sensitive) environment variables from `Values` config, referring
        # them from the related Secret object. This is an extra measure of security
        # to avoid having credentials displayed in plain text in the template file.
        # https://kubernetes.io/docs/concepts/configuration/secret/
        {{- $chartName := .Chart.Name -}}
        {{- range $key, $val := .Values.secrets }}
        - name: {{ $key }}
          valueFrom:
            secretKeyRef:
              name: {{ $chartName }}-secret
              key: {{ $key }}
          {{- end}}
```

File: incubator/prisma/values.yaml
Comment (emmenko) [reply incubator/prisma/values.yaml]: Well, the thing is that installing the postgres/mysql DB charts should be optional. It's required if you don't have any installed yet, but if you already have a running DB you don't need the extra chart.
And yeah, prisma will eventually crash if it can't find/connect to the DB. But that's fine 😉

File: incubator/prisma/values.yaml
Comment (emmenko) [reply incubator/prisma/values.yaml]: I see, it would be better to explicitly mark it otherwise it will default to `latest` if not specified

File: incubator/prisma/values.yaml
Comment (emmenko) [reply incubator/prisma/values.yaml]: Yeah, although I'm thinking that we might end up with a complex chart if we start supporting all the connectors.

Based also on my [other comment](https://github.com/kubernetes/charts/pull/6483#discussion_r203616442) I would actually suggest to remove the DB chart dependency. I think it's fine to just document what to do. At least as a first step. We could always decide to add it later on if we see a need for that.

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: A database is an intrinsic requirement for Prisma. They can't be separated. Users don't use Prisma because they have a database. They set up a database because they need to use Prisma. 
Also, I don't think the chart will pass CI going into CrashLoop state, which is the behaviour when a database connector is missing.
This chart has been modeled after the [Kafka](https://github.com/helm/charts/tree/master/incubator/kafka) one. It installs with dependencies by default, but it also lets you use your own.

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: Actually, recent experience has taught me that it is not a good idea to change an image's tag since the chart composition might depend on it. In other words, charts don't work with different versions of an image (most of the times).

Also, it doesn't default to `latest`, but to `9.6.2` since that is the tag of the `0.15.0` version of the PostgreSQL chart.

File: incubator/prisma/values.yaml
Comment (giacope) [reply incubator/prisma/values.yaml]: The existence of the PostgreSQL chart is just for dependency purposes in the same way Zookeeper is to [Kafka](https://github.com/helm/charts/tree/master/incubator/kafka).

I think it's also not a good idea to support all connectors. But it's essential to at least have one set up to ease quickstarts. We don't want users to run into complex configurations by default, but just leave that as an option for those who want to take a step further.

File: incubator/prisma/templates/configmap.yaml
Comment (giacope) [reply incubator/prisma/templates/configmap.yaml]: Looks good 👍 I will add it soon. Thanks!

File: incubator/prisma/requirements.yaml
Comment (mattfarina) [inline incubator/prisma/requirements.yaml]: Should this be a version range instead of a flat version? For example, ~0.15.0 which is >= 0.15.0, < 0.16.0

File: incubator/prisma/templates/configmap.yaml
Comment (giacope) [reply incubator/prisma/templates/configmap.yaml]: @emmenko Actually, it's not possible through this method because `host: localhost` is a value that depends on the release name, so it can't be something in the `values.yaml` file. It can also lead to many configuration errors. IMHO, the best way is that Prisma supports configuring sensitive information through environment variables.

File: incubator/prisma/requirements.yaml
Comment (giacope) [reply incubator/prisma/requirements.yaml]: Interesting! Just added that.

File: incubator/prisma/values.yaml
Comment (TD-4242) [reply incubator/prisma/values.yaml]: perhaps a simple non-persistent db deployment for testing and CI purposes baked into this chart that can easily be turned off to point to a real deployment.

Comment (giacope) [reply incubator/prisma/values.yaml]: I mean, the PostgreSQL dependency is a real deployment. But anyways, I'll set `persistence.enabled` to `false` to make it simpler.

**Comment by cpanato** [comment]
@giacomoguiulfo please add the `OWNERS` file in the `.helmignore` - https://github.com/helm/charts#owning-and-maintaining-a-chart

File: incubator/prisma/templates/deployment.yaml
Comment (cpanato) [inline incubator/prisma/templates/deployment.yaml]: can we add a `Liveness and Readiness Probes` for the pod? or the app does not have any health check endpoint?

Comment (giacope) [reply incubator/prisma/templates/deployment.yaml]: The app doesn't have a (documented) health check endpoint.

**Comment by giacomoguiulfo** [comment]
@cpanato Done 👍

**Review by plmercereau — CHANGES_REQUESTED** [review]
Hello,
I think there is a type in your ingress.yaml: your backend servicePort is http where it should be {{ $servicePort }}
Best

**Comment by giacomoguiulfo** [comment]
Good catch @plmercereau, thanks!

**Comment by unguiculus** [comment]
I'd suggest you move the chart to stable.

**Comment by giacomoguiulfo** [comment]
Thanks, @unguiculus. Just moved the chart on this PR to `stable` 🚀

**Comment by noelebrun** [comment]
Hello, is something still missing in this PR to be merged? Can I help in some way?

**Comment by giacomoguiulfo** [comment]
Hello @noelebrun. For this PR to be merged the only thing missing is for a `Helm/charts` maintainer to add the `lgtm` label. This should happen shortly. Thank you! 😀

File: stable/prisma/README.md
Comment (unguiculus) [inline stable/prisma/README.md]: nil -> ""

File: stable/prisma/README.md
Comment (unguiculus) [inline stable/prisma/README.md]: nil -> ""

File: stable/prisma/templates/deployment.yaml
Comment (unguiculus) [inline stable/prisma/templates/deployment.yaml]: Can you use apps/v1?

File: stable/prisma/ci/example-values.yaml
Comment (unguiculus) [inline stable/prisma/ci/example-values.yaml]: The file should only contain changes over the default values.yaml. Don't copy everything.

File: stable/prisma/values.yaml
Comment (unguiculus) [inline stable/prisma/values.yaml]: resources: {}

File: stable/prisma/README.md
Comment (unguiculus) [inline stable/prisma/README.md]: Update all references to incubator.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/prisma/ci/example-values.yaml
Comment (agonbina) [inline stable/prisma/ci/example-values.yaml]: Are you aware of this issue [here](https://github.com/prisma/prisma/issues/2850) re: scaling Prisma to multiple instances? I have not looked at your chart implementation and architecture design decisions as I do not have the best understanding of Kubernetes, but just wanted to make you aware of the specifics required to run Prisma in a horizontally scalable architecture.

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [inline stable/prisma/templates/deployment.yaml]: Is this a reference to ".Values.auth.secret"?

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [inline stable/prisma/templates/deployment.yaml]: Is this not used as "PRISMA_SECRET" in the ConfigMap?

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [inline stable/prisma/templates/deployment.yaml:46]: Where is this file created? Did I missed something?

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [inline stable/prisma/templates/deployment.yaml:33]: I'm not sure Prisma image is using a CONFIG environment variable? Isn't it "PRISMA_CONFIG"?

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [inline stable/prisma/templates/deployment.yaml]: Should you not use the password generated in Postgres Secret if ".Values.postgresql.enabled" is true?

File: stable/prisma/templates/secret.yaml
Comment (noelebrun) [inline stable/prisma/templates/secret.yaml]: You should probably let Postgres chart handle the password is Postgresql is enabled

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml]: Yes, good catch!

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml]: Yes, my bad. Good catch again!

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml:46]: It's created by the `prerun_hook.sh` script that gets executed before the Prisma server starts.

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml:33]: The `CONFIG` environment variable is used by the `prerun_hook.sh` to `envsubst` all the env variables passed.

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml]: Good idea! Done

File: stable/prisma/templates/secret.yaml
Comment (giacope) [reply stable/prisma/templates/secret.yaml]: 👌

File: stable/prisma/templates/deployment.yaml
Comment (giacope) [reply stable/prisma/templates/deployment.yaml]: Sure

File: stable/prisma/ci/example-values.yaml
Comment (giacope) [reply stable/prisma/ci/example-values.yaml]: Hello @agonbina.

I wasn't aware of this issue. I agree that this is something the should definitely be addressed by the chart, but I will prefer it not to be now since it requires major modifications (RabbitMQ server dependency, additional deployment for the "primary" Prisma server, and so on). I added a note in the README saying that this chart does not support horizontal scaling for Prisma yet.

The reason why I prefer to don't make the necessary changes for horizontal scalability now is that this PR is close to being 3 months old, and it's hard to have multiple contributors to the chart since the chart doesn't really "exists" on the repo. I will like for this chart to go upstream first so that more people can contribute to it and perhaps tackle the issue you just mentioned (and maybe some others) much faster.

What do you think? 🙂

File: stable/prisma/templates/deployment.yaml
Comment (noelebrun) [reply stable/prisma/templates/deployment.yaml:46]: Oh, yes you're using the heroku version. Is this version documented somewhere?

Comment (giacope) [reply stable/prisma/templates/deployment.yaml:46]: I don't think so :(

Comment (noelebrun) [reply stable/prisma/templates/deployment.yaml:46]: Oh okay, I was curious to see how they implemented the heroku one. You find it in Prisma sources?

Comment (giacope) [reply stable/prisma/templates/deployment.yaml:46]: I heard about it from a Prisma open source user. Then I dig in around the container by myself to figure it out.

File: stable/prisma/ci/example-values.yaml
Comment (noelebrun) [reply stable/prisma/ci/example-values.yaml]: Jumping into the conversation 😄 I would like to help on this one, once this first version will be merged. Be careful tho that you for now you need to use the prisma-prod image for this, which doesn't have the `envsubst` hook the manage your config.

So, you will probably have to either build your own image with a custom similar hook, or wait for [this issue](https://github.com/prisma/prisma/issues/2869) to be solved and hope that the merged version will have an heroku version.

Comment (giacope) [reply stable/prisma/ci/example-values.yaml]: Yeah, you are right. Once this is in, and there is something "tangible", we will be able to tackle issues faster and make the chart better by having more contributors (including Prisma people perhaps). Everything will be sorted out with time.

**Comment by giacomoguiulfo** [comment]
ping @unguiculus

**Review by unguiculus — APPROVED** [review]

**Pull request merged**
